crates/node-maintainer/src/maintainer.rs

```diff
@@ -173,14 +178,40 @@ impl NodeMaintainer {
         let (package_sink, package_stream) = futures::channel::mpsc::unbounded();
         let mut q = VecDeque::new();
         q.push_back(self.graph.root);
+
+        // Number of dependencies queued for processing in `package_stream`
         let mut in_flight = 0;
 
+        // Since we queue dependencies for multiple packages at once - it is
+        // not unlikely that some of them would be duplicated by currently
+        // fetched dependencies. Thus we maintain a mapping from "name@spec" to
+        // a vector of `NodeDependency`s. When we will fetch the package - we
+        // will apply it to all dependencies that need it.
+        let fetches: HashMap<String, Vec<NodeDependency>> = HashMap::new();
+        let fetches = Arc::new(Mutex::new(fetches));
+
         let mut package_stream = package_stream
             .map(|dep: NodeDependency| {
-                self.nassun
-                    .resolve(format!("{}@{}", dep.name, dep.spec))
-                    .map_ok(move |p| (p, dep))
+                let spec = format!("{}@{}", dep.name, dep.spec);
+                let maybe_spec = if let Some(mut fetches) = fetches.try_lock() {
+                    if let Some(list) = fetches.get_mut(&spec) {
+                        // Package fetch is already in-flight, add dependency
+                        // to the existing list.
+                        list.push(dep);
+                        None
+                    } else {
+                        // Fetch package since we are the first one to get here.
+                        fetches.insert(spec.clone(), vec![dep]);
+                        Some(spec)
+                    }
+                } else {
+                    // Mutex is locked - fetch the package
+                    Some(spec)
```

Comment:
So technically this should never ever happen, but if we ever switch to a truly concurrent model - I don't want this code to throw.